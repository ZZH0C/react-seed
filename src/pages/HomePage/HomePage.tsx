import React from 'react';

import { Head } from '../../components/Head/Head';
import { Navbar } from '../../components/Navbar/Navbar';
import { MainContainer } from '../../components/MainContainer/MainContainer';
import { NavbarItem } from '../../components/NavbarItem/NavbarItem';
import { HeaderLogo } from '../../components/HeaderLogo/HeaderLogo';
import { HeaderUser } from '../../components/HeaderUser/HeaderUser';
import { SubmenuItem } from '../../components/Submenu/SubmenuItem/SubmenuItem';
import { testProps_logo } from '../../components/mockProps';
import { SubmenuSubItem } from '../../components/Submenu/SubmenuSubItem/SubmenuSubItem';
import { HeaderUserIcons } from '../../components/HeaderUserIcons/HeaderUserIcons';
import { Submenu } from '../../components/Submenu/Submenu';
import { MainSection } from '../../components/MainSection/MainSection';
import {
  GoogleLogin,
  GoogleLoginResponse,
  GoogleLoginResponseOffline,
  GoogleLogout,
} from 'react-google-login';
import { useUserContext } from '../../hooks/useUserContext';
import { useUserData } from '../../hooks/useUserData';
import {
  clientId,
  emptyProfilePictureSrc,
  SubItemCategories,
  categoryContainer,
  navBarLabels,
} from '../../components/config';
export const UserContext = useUserContext;
export const HomePage: React.FC = () => {
  const { state, dispatch } = useUserData();

  const responseGoogle = (
    response: GoogleLoginResponse | GoogleLoginResponseOffline,
  ) => {
    if ('profileObj' in response) {
      dispatch({ type: 'logIn', user: response });
    }
  };
  const responseFailure = (resp: { error: string }) => {
    console.log(resp);
    // TODO: add error handler
    throw new Error(resp.error);
  };
  const logout = () => {
    dispatch({ type: 'logOut' });
  };

  const subItems = SubItemCategories.map((subItemProps) => {
    return (
      <SubmenuSubItem
        key={Math.random()}
        href={subItemProps.href}
        name={subItemProps.name}
        category={subItemProps.category}
        // isActive={false}
      >
        {subItemProps.name}
      </SubmenuSubItem>
    );
  });
  const navBarItems = navBarLabels.map((navBarItemProps) => {
    return (
      <NavbarItem
        key={Math.random()}
        href={navBarItemProps.href}
        name={navBarItemProps.name}
        label={navBarItemProps.label}
      >
        {navBarItemProps.name}
      </NavbarItem>
    );
  });

  return (
    <section>
      <UserContext.Provider value={state.userData}>
        <GoogleLogin
          clientId={clientId}
          buttonText="Login"
          onSuccess={responseGoogle}
          onFailure={responseFailure}
          disabled={state.isLogged}
          disabledStyle={{ display: 'none' }}
        />
        <GoogleLogout
          clientId={clientId}
          buttonText="Logout"
          onLogoutSuccess={logout}
          disabled={!state.isLogged}
          disabledStyle={{ display: 'none' }}
        />
        <Head>
          <HeaderLogo
            mobileImgSrc={testProps_logo.mobileImgSrc}
            mobileImgAlc={testProps_logo.mobileImgAlc}
            desktopImgSrc={testProps_logo.desktopImgSrc}
            desktopImgAlt={testProps_logo.desktopImgAlt}
          >
            Page Title
          </HeaderLogo>
          <HeaderUser src={emptyProfilePictureSrc}>
            <HeaderUserIcons classname={'icon-bell'} />
            <HeaderUserIcons classname={'icon-mail'} />
          </HeaderUser>
        </Head>
        <Navbar>{navBarItems}</Navbar>
        <MainContainer>
          <Submenu>
            <SubmenuItem name={categoryContainer.name}>{subItems}</SubmenuItem>
          </Submenu>
          <MainSection />
        </MainContainer>
      </UserContext.Provider>
    </section>
  );
};
