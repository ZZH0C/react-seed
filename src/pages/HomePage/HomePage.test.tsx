import { configure, shallow } from 'enzyme';
import { HomePage } from './HomePage';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

configure({ adapter: new Adapter() });

jest.mock('../../hooks/useUserContext/useUserContext');

jest.mock('../../components/config', () => {
  const fakeItems = {
    href: '/',
    name: 'FAKE_NAME',
    category: 'FAKE_CATEGORY',
  };
  const fakeLabel = {
    href: '/',
    name: 'FAKE_NAME',
    label: 'FAKE_LABEL',
  };
  const fakeSrc = '/';
  const fakeCategoryContainer = { name: 'FAKE_CATEGORY_NAME' };
  const fakeLogoProps = {
    desktopImgSrc: 'FAKE_SRC',
    mobileImgSrc: 'FAKE_SRC',
    desktopImgAlt: 'FAKE_SRC',
    mobileImgAlc: 'FAKE_SRC',
  };

  return {
    SubItemCategories: [fakeItems, fakeItems],
    navBarLabels: [fakeLabel, fakeLabel],
    emptyProfilePictureSrc: fakeSrc,
    categoryContainer: fakeCategoryContainer,
    logoProps: fakeLogoProps,
  };
});

describe('components/HomePage', () => {
  it('should render', () => {
    expect(shallow(<HomePage />)).toMatchSnapshot();
  });
});
