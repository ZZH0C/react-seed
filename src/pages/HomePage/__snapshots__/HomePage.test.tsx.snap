// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/HomePage should render 1`] = `
<section>
  <Head>
    <HeaderLogo
      desktopImgAlt="FAKE_SRC"
      desktopImgSrc="FAKE_SRC"
      mobileImgAlc="FAKE_SRC"
      mobileImgSrc="FAKE_SRC"
    >
      Page Title
    </HeaderLogo>
    <HeaderUser
      src="/"
    >
      <HeaderUserIcons
        classname="icon-bell"
      />
      <HeaderUserIcons
        classname="icon-mail"
      />
    </HeaderUser>
  </Head>
  <Navbar>
    <NavbarItem
      href="/"
      key="0"
      label="FAKE_LABEL"
      name="FAKE_NAME"
    >
      FAKE_NAME
    </NavbarItem>
    <NavbarItem
      href="/"
      key="1"
      label="FAKE_LABEL"
      name="FAKE_NAME"
    >
      FAKE_NAME
    </NavbarItem>
  </Navbar>
  <MainContainer>
    <Submenu>
      <SubmenuItem
        name="FAKE_CATEGORY_NAME"
      >
        <SubmenuSubItem
          category="FAKE_CATEGORY"
          href="/"
          key="0"
          name="FAKE_NAME"
        >
          FAKE_NAME
        </SubmenuSubItem>
        <SubmenuSubItem
          category="FAKE_CATEGORY"
          href="/"
          key="1"
          name="FAKE_NAME"
        >
          FAKE_NAME
        </SubmenuSubItem>
      </SubmenuItem>
    </Submenu>
    <MainSection />
  </MainContainer>
</section>
`;
